src/routing_table/mod.rs

```diff
@@ -499,68 +464,63 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
+            let merge_prefix = self.our_group_prefix.popped();
+            let mut groups = self.groups.clone();
+            // Safe to unwrap as `self.groups` must always contain `our_group_prefix`.
+            let _ = unwrap!(groups.get_mut(&self.our_group_prefix)).insert(self.our_name);
+            removal_details.targets_and_merge_details =
+                Some((self.prefixes_within_merge(&merge_prefix),
+                      OwnMergeDetails {
+                    sender_prefix: self.our_group_prefix,
+                    merge_prefix: merge_prefix,
+                    groups: groups,
+                }));
         }
         Ok(removal_details)
     }
 
-    // Merges our own group and all existing compatible groups into the new one defined by
-    // `merge_details.prefix`.
+    // When a merge of our own group is triggered (either from our own group or a neighbouring one)
+    // this function handles the incoming merge details from the peers within the merging groups.
     //
-    // The appropriate targets (all contacts which are not part of the merging groups) and the merge
-    // details they each need to receive (the new prefix and the new group) is returned.
-    pub fn merge_own_group(&mut self,
-                           merge_details: &OwnMergeDetails<T>)
-                           -> (Vec<T>, OtherMergeDetails<T>) {
-        self.merge(&merge_details.prefix);
-
-        // For each provided group which is not currently in our routing table and which is not one
-        // of the merging groups, add an empty group and cache the corresponding contacts
-        let mut needed = HashSet::<T>::new();
-        for (prefix, contacts) in merge_details.groups
-            .iter()
-            .filter(|&(prefix, _)| !merge_details.prefix.is_compatible(prefix)) {
+    // The actual merge of the group is only done once all expected merging groups have provided
+    // details.  See the docs for `OwnMergeState` for full details of the return value.
+    pub fn merge_own_group(&mut self, merge_details: OwnMergeDetails<T>) -> OwnMergeState<T> {
+        for (prefix, contacts) in &merge_details.groups {
+            // Cache the prefix if it's a merging group, flagging the sender's prefix `true`.
+            if merge_details.merge_prefix.is_compatible(prefix) {
+                let prefix_entry = self.merging.entry(*prefix).or_insert(false);
+                if merge_details.sender_prefix == *prefix {
+                    *prefix_entry = true;
+                }
+            }
+            // Add an empty group in the table and cache the corresponding contacts.
             if self.groups.entry(*prefix).or_insert_with(HashSet::new).is_empty() {
-                self.needed.extend(contacts.iter());
-                needed.extend(contacts.into_iter());
+                let needed_group = self.needed.entry(*prefix).or_insert_with(HashSet::new);
+                *needed_group = needed_group.union(contacts).cloned().collect();
```

Comment:
Yup - it would.  In theory there shouldn't be much difference performance-wise since this _should_ always yield `needed_group` as empty or as identical to `contacts`, but `extend()` is certainly simpler.